### PR TITLE
Add hot-swappable EXL3 MSRT expert cartridges with TP support

### PR DESCRIPTION
**Pull request by malaiwah**
## Summary

This adds opt-in, model-wide hot-swapping of packed EXL3 MSRT residual cartridges for routed-expert MoE models.

Operators can switch additive expert-quality stages without materializing dense expert weights or restarting the engine. `AsyncLLM` stages and validates the cartridge before pausing serving, drains requests, replaces the fixed-address CUDA-graph state, recaptures, and resumes. Failed transitions restore the compressed base path.

## Artifact contracts

This PR implements the closed cross-repository runtime contracts:

- base checkpoint profile: `exl3-msrt-base/1`
- adapter schema: `fq-cartridge-adapter/3`
- adapter runtime profile: `exl3-msrt-additive/1`
- ExLlamaV3 extension ABI: `EXL3_MOE_ADDITIVE_ABI_VERSION = 1`

The loader validates ordered residual chains, parent-closed sparse coverage, exact layer/expert/projection/stage coverage, uniform base bitrate, scalar FP32 scales, MCG semantics, tensor geometry, and the base's per-layer byte identities plus aggregate root.

## Tensor parallelism

TP is supported end to end, including TP > 1.

- Full-rank base and adapter artifacts use logical `rank0` tensors and are sliced consistently at runtime.
- Rank-sharded adapters declare an exact world size and rank set; rank-sharded TP=1 remains distinct from full-rank storage.
- Replicated stage scales must match bit-for-bit across ranks.
- Base compatibility identities are computed from full logical tensor bytes before slicing, so identity is invariant across supported runtime TP layouts.

## Runtime lifecycle and safety

Dynamic loading requires an independent operator opt-in:

```bash
VLLM_ENABLE_EXL3_CARTRIDGE=1
```

Checkpoint metadata advertises capability but cannot enable this control path.

Before serving is paused, each worker validates the closed manifest and extension ABI, copies only bounded regular files into private staging while checking declared size and SHA-256, and preflights coverage, topology, scales, K values, and tensor geometry. The drained transaction then performs only model materialization and CUDA-graph replacement.

The implementation also provides:

- one mutation lock shared with pause/resume controls;
- bounded distributed staging RPC and uniform worker-count checks;
- shielded rollback, staging cleanup, and resume paths;
- cleanup after partial-worker failures and correct A-to-narrower-B replacement;
- model-scoped fixed-address scratch shared across sequential MoE layers;
- generic development-API errors while retaining server-side diagnostics.

The API accepts only trusted-admin, worker-local paths. `producer_verified_signer` is provenance metadata, not runtime authentication.

## Current scope

Supported initially:

- V1 `AsyncLLM`;
- tensor parallelism, including TP > 1;
- one model-wide active cartridge;
- the MCG codebook with base-owned rotations.

Explicitly rejected for now:

- synchronous `LLMEngine` hot-swapping;
- pipeline, data, or expert parallelism;
- the V2 model runner;
- simultaneous vLLM LoRA adapters.

## Companion changes

This is one coordinated three-repository implementation:

- [ExLlamaV3 #284](https://github.com/turboderp-org/exllamav3/pull/284): fused additive MoE kernel and versioned ABI.
- [progressive-tensors #41](https://github.com/malaiwah/progressive-tensors/pull/41): base/cartridge production, schemas, TP layouts, and compatibility identities.

## Validation

- 93/93 focused and review-regression tests passed on a two-GPU Blackwell host.
- A real two-process NCCL TP=2 run exercised `Exl3MoEMethod.create_weights` and the full-rank loading path on both GPUs; gathered gate/up/down trellises and rotations reconstructed the original tensors exactly, and both ranks verified the same pre-slice byte identity.
- Unit coverage includes full and rank-sharded adapters (including rank-sharded TP=1), malformed rank topology, divergent scale replicas, parent closure, source replacement, bounded-file handling, wrapped live-layer names, narrower swaps, compile warmups, malformed UTF-8 requests, and the Async failure/cancellation matrix.
- Producer and consumer assert the same hard-coded compatibility golden: layer `3518095a...377d58`, root `cbb1f591...2ce01`.
- Local Ruff formatting/lint, Python compilation, and diff-whitespace checks pass.
- Earlier RTX 5090 functional validation covered base → active cartridge → base graph recapture with exact restored-base logprobs (`max |delta| = 0.0`).

## AI assistance

AI assistance was used for implementation and adversarial review. I reviewed the changed code, focused tests, cross-repository contracts, and GPU validation evidence.

### CODE REVIEW

**Comment by malaiwah** [comment]
Reviewed hardening is pushed to the head repository branch as `d057cbc05a` (`malaiwah/vllm-voipmonitor:feat/exl3-lora-cartridge-clean`).

The GitHub PR currently has `maintainer_can_modify=false`, so the source-repository push does not update this PR automatically; please enable maintainer edits or refresh/reopen the PR from the fork branch.

Verification performed on an RTX 5090 with source overlays (no image rebuild):
- Focused cartridge suite: `28 passed`
- Full vLLM CUDA-graph hot-swap: K2 base -> K3-like -> K3+K4-like -> base; 10/10 routed MoE layers updated at each transition; restored base logprobs exactly (`max_abs=0.0`)
- Mean KLD: K2->K3-like `0.268769`; K3-like->K3+K4-like `0.119068`; K2->K3+K4-like `0.352634`
- Captured-decode benchmark, 2,048 output tokens: runtime disabled `4515.8 tok/s`; inactive runtime before null routing `4047.4 tok/s`; after graph-stable null routing `4094.9 tok/s`
- `ruff check` passed on all modified production/test files

Adversarial correctness/CUDA review found no remaining blocker/high correctness or graph-capture defect after fixes. AI assistance was used; the human submitter must review every changed line and reproduce relevant tests before merge.

**Comment by malaiwah** [comment]
Correction: PR #299 has updated successfully to head `d057cbc05a`. The earlier note about needing maintainer edits/reopening was incorrect; no action is required to refresh the PR.

**Comment by malaiwah** [comment]
Final current-head audit for `e1610198c7` (RTX 5090, `cudagraph_mode=FULL`, exact source overlays):

- Runtime disabled: **4514.900 output tok/s**, **31,194,349,568 device bytes** after initialization.
- Runtime enabled but inactive: **4521.333 output tok/s**, **31,194,349,568 device bytes** after initialization.
- Measured inactive throughput delta: **+0.1425%** (noise-level); inactive VRAM delta: **0 bytes**.
- Active K3-like cartridge: **30,118,510,592 device bytes** versus **22,596,026,368** immediately-before-load base state; dense allocation appears only while active.
- After unload: **22,065,446,912 device bytes**, **506 MiB below** the immediately-before-load base allocator state.
- Full K2 -> K3-like -> K3+K4-like -> base transition updated/deactivated **10/10 layers** and restored base logits exactly (`max_abs=0.0`).
- Fresh local suite: **41 passed, 1 skipped**. Fresh RTX 5090 suite: **42 passed**. Ruff and Python compilation passed.

This directly verifies that enabling cartridge support without loading a cartridge has no measurable throughput penalty and no VRAM increase in this workload.

**Comment by malaiwah** [comment]
Packed MSRT runtime update pushed in `4abd74542c`.

The cartridge path now retains EXL3 trellises and executes base plus residual projections through the proposed upstream ExLlamaV3 fused API instead of materializing dense BF16 expert shadows. Upstream dependency: turboderp-org/exllamav3#284.

Validation on RTX 5090, GLM-5.2 K2, full CUDA-graph capture:

- K2 base -> packed K3-like cartridge -> base transition completed
- deactivation restored base logprobs exactly (`max_abs=0.0`)
- packed vs prior dense cartridge reference: mean KL `1.30506e-4`, max absolute logit delta `0.11122`, mean absolute logit delta `0.00910`, sampled argmax identical
- cartridge file: `583,794,400` bytes
- measured active runtime delta in one isolated run: `161,480,704` device bytes (allocator measurements vary across graph recapture because CUDA graph pools are released/rebuilt)
- steady-state sampled throughput was run-order sensitive: packed/base ratio ranged from `0.783` to `1.140`; no stable regression claim from these short runs
- packed cartridge load plus graph recapture: 142-197 s; unload plus base recapture: ~1.0 s

Focused local checks:

```text
uvx ruff check ...
All checks passed

.venv/bin/python -m pytest tests/quantization/test_exl3_lora_cartridge.py -q
46 passed, 14 warnings in 4.51s
```

This update also rejects non-finite or FP32-overflowing inverse scales before activation and keeps the base path compressed after cartridge unload.

**Comment by malaiwah** [comment]
Final dependency verification: ExLlamaV3 PR #284 follow-up `4d00db7` was rebuilt and rerun through this branch. Full CUDA-graph K2 base -> packed cartridge -> base passed: 10 layers activated/deactivated, active device delta `157,286,400` bytes, load+recapture `112.05 s`, unload+recapture `1.004 s`, active-vs-base max logprob delta `8.6530676`, restored-vs-base max delta `0.0`. Short-run packed/base throughput ratio was `0.7694` in this concurrent-GPU run; the PR body retains the broader observed run-order range and makes no stable performance claim.

**Comment by malaiwah** [comment]
Maintainer note: this PR is ready for review, but fork authors cannot apply the repository `ready` label. The current red `pre-run-check` is only that policy gate (author has 1 merged PR); the actual pre-commit job is skipped. Please apply `ready` or `verified` when appropriate to run CI. Local/remote validation is documented in the PR body, including 93/93 focused and review-regression tests on a 2×Blackwell host and a real two-process NCCL TP=2 reconstruction run.
